File: backend-node/src/services/settings.service.js
import '../load-env.js';
import { supabaseAdmin } from './supabase.service.js';

const PAYMENT_PROVIDER_KEY = 'payment_provider';
const VALID_PROVIDERS = ['stripe', 'mercadopago'];

let memoryCache = null;

const normalizeProvider = (value) => {
  const v = String(value || '').trim().toLowerCase();
  return VALID_PROVIDERS.includes(v) ? v : 'stripe';
};

const getEnvDefault = () => normalizeProvider(process.env.PAYMENT_PROVIDER || 'stripe');

/**
 * Obtém o provedor de pagamento ativo (stripe | mercadopago)
 */
export const getPaymentProvider = async () => {
  if (memoryCache) return memoryCache;

  if (supabaseAdmin) {
    try {
      const { data, error } = await supabaseAdmin
        .from('app_configuracoes')
        .select('valor')
        .eq('chave', PAYMENT_PROVIDER_KEY)
        .maybeSingle();

      if (!error && data?.valor) {
        memoryCache = normalizeProvider(data.valor);
        return memoryCache;
      }
    } catch (err) {
      console.warn('⚠️ Tabela app_configuracoes indisponível, usando PAYMENT_PROVIDER do .env:', err.message);
    }
  }

  memoryCache = getEnvDefault();
  return memoryCache;
};

/**
 * Define o provedor de pagamento (admin)
 */
export const setPaymentProvider = async (provider) => {
  const normalized = normalizeProvider(provider);
  memoryCache = normalized;

  if (!supabaseAdmin) {
    console.warn('⚠️ Supabase não configurado — provedor salvo apenas em memória até reiniciar o servidor.');
    return normalized;
  }

  const now = new Date().toISOString();
  const { error } = await supabaseAdmin
    .from('app_configuracoes')
    .upsert(
      { chave: PAYMENT_PROVIDER_KEY, valor: normalized, atualizado_em: now },
      { onConflict: 'chave' }
    );

  if (error) {
    memoryCache = null;
    throw new Error(`Erro ao salvar provedor de pagamento: ${error.message}`);
  }

  return normalized;
};

export const getValidPaymentProviders = () => [...VALID_PROVIDERS];

export const clearPaymentProviderCache = () => {
  memoryCache = null;
};

File: backend/src/CurriculosProIA.Service/Implementations/SettingsService.cs
using CurriculosProIA.Service.Interfaces;
using CurriculosProIA.Repository.Interfaces;
using CurriculosProIA.Service.Helpers;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.Logging;

namespace CurriculosProIA.Service.Implementations;

public class SettingsService : ISettingsService
{
    private const string PaymentProviderKey = "payment_provider";
    private const string MercadoPagoModeKey = "mercadopago_mode";
    private static readonly string[] ValidProviders = ["stripe", "mercadopago", "cakto", "kiwify"];
    private static readonly string[] ValidMercadoPagoModes =
        [MercadoPagoConfigHelper.ModeTest, MercadoPagoConfigHelper.ModeProduction];

    private readonly IAppSettingsRepository _settingsRepo;
    private readonly IConfiguration _configuration;
    private readonly ILogger<SettingsService> _logger;

    public SettingsService(
        IAppSettingsRepository settingsRepo,
        IConfiguration configuration,
        ILogger<SettingsService> logger)
    {
        _settingsRepo = settingsRepo;
        _configuration = configuration;
        _logger = logger;
    }

    public IReadOnlyList<string> GetValidPaymentProviders() => ValidProviders;

    public void ClearPaymentProviderCache()
    {
    }

    public async Task<string> GetPaymentProviderAsync(CancellationToken cancellationToken = default)
    {
        try
        {
            var value = await _settingsRepo.GetAppConfigValueAsync(PaymentProviderKey, cancellationToken);
            if (!string.IsNullOrEmpty(value))
            {
                return NormalizeProvider(value);
            }
        }
        catch (Exception ex)
        {
            _logger.LogWarning(ex, "Tabela app_configuracoes indisponível, usando PAYMENT_PROVIDER do .env");
        }

        return GetEnvDefault();
    }

    public async Task<string> SetPaymentProviderAsync(string provider, CancellationToken cancellationToken = default)
    {
        var normalized = NormalizeProvider(provider);

        try
        {
            await _settingsRepo.SetAppConfigValueAsync(PaymentProviderKey, normalized, cancellationToken);
        }
        catch (Exception ex)
        {
            throw new InvalidOperationException($"Erro ao salvar provedor de pagamento: {ex.Message}", ex);
        }

        return normalized;
    }

    public IReadOnlyList<string> GetValidMercadoPagoModes() => ValidMercadoPagoModes;

    public void ClearMercadoPagoModeCache()
    {
    }

    public async Task<string> GetMercadoPagoModeAsync(CancellationToken cancellationToken = default)
    {
        try
        {
            var value = await _settingsRepo.GetAppConfigValueAsync(MercadoPagoModeKey, cancellationToken);
            if (!string.IsNullOrEmpty(value))
            {
                return NormalizeMercadoPagoMode(value);
            }
        }
        catch (Exception ex)
        {
            _logger.LogWarning(ex, "Tabela app_configuracoes indisponível, usando MERCADOPAGO_MODE do .env");
        }

        return MercadoPagoConfigHelper.GetMode(_configuration);
    }

    public async Task<string> SetMercadoPagoModeAsync(string mode, CancellationToken cancellationToken = default)
    {
        var normalized = NormalizeMercadoPagoMode(mode);

        try
        {
            await _settingsRepo.SetAppConfigValueAsync(MercadoPagoModeKey, normalized, cancellationToken);
        }
        catch (Exception ex)
        {
            throw new InvalidOperationException($"Erro ao salvar ambiente Mercado Pago: {ex.Message}", ex);
        }

        return normalized;
    }

    private string GetEnvDefault() =>
        NormalizeProvider(_configuration["PAYMENT_PROVIDER"] ?? "stripe");

    private static string NormalizeProvider(string? value)
    {
        var v = (value ?? string.Empty).Trim().ToLowerInvariant();
        return ValidProviders.Contains(v) ? v : "stripe";
    }

    private static string NormalizeMercadoPagoMode(string? value)
    {
        var v = (value ?? string.Empty).Trim().ToLowerInvariant();
        return ValidMercadoPagoModes.Contains(v) ? v : MercadoPagoConfigHelper.ModeTest;
    }
}
